posts/2021-04-30-edition-2021.md

```diff
@@ -0,0 +1,350 @@
+It's identical to the current one, except for three new additions:
+
+- [`std::convert::TryInto`](https://doc.rust-lang.org/stable/std/convert/trait.TryInto.html)
+- [`std::convert::TryFrom`](https://doc.rust-lang.org/stable/std/convert/trait.TryFrom.html)
+- [`std::iter::FromIterator`](https://doc.rust-lang.org/stable/std/iter/trait.FromIterator.html)
+
+The library team still needs to formally approve these, which will likely happen soon.
+<!-- TODO: hopefully this happens before we publish this -->
+
+### Default Cargo feature resolver
+
+Since Rust 1.51.0, Cargo has opt-in support for a [new feature resolver][4]
+which can be activated with `resolver = "2"` in `Cargo.toml`.
+
+Starting in Rust 2021, this will be the default.
+That is, writing `edition = "2021"` in `Cargo.toml` will imply `resolver = "2"`.
+
+The new feature resolver no longer merges all requested features for
+crates that are depended on in multiple ways.
+See [the announcement of Rust 1.51][5] for details.
+
+[4]: https://doc.rust-lang.org/cargo/reference/resolver.html#feature-resolver-version-2
+[5]: https://blog.rust-lang.org/2021/03/25/Rust-1.51.0.html#cargos-new-feature-resolver
+
+### IntoIterator for arrays
+
+Until Rust 1.53, only *references* to arrays implement `IntoIterator`.
+This means you can iterate over `&[1, 2, 3]` and `&mut [1, 2, 3]`,
+but not over `[1, 2, 3]` directly.
+
+```rust
+for &e in &[1, 2, 3] {} // Ok :)
+
+for e in [1, 2, 3] {} // Error :(
+```
+
+This has been [a long-standing issue][25], but the solution is not as simple as it seems.
+Just [adding the trait implementation][20] would break existing code.
+`array.into_iter()` already compiles today because that implicitly calls
+`(&array).into_iter()` due to [how method call syntax works][22].
+Adding the trait implementation would change the meaning.
+
+Usually we categorize this type of breakage
+(adding a trait implementation) 'minor' and acceptable.
+But in this case there is too much code that would be broken by it.
+
+It has been suggested many times to "only implement `IntoIterator` for arrays in Rust 2021".
+However, this is simply not possible.
+You can't have a trait implementation exist in one edition and not in another,
+since editions can be mixed.
+
+Instead, we decided to add the trait implementation in all editions (starting in Rust 1.53.0),
+but add a small hack to avoid breakage until Rust 2021.
+In Rust 2015 and 2018 code, the compiler will still resolve `array.into_iter()`
+to `(&array).into_iter()` like before, as if the trait implementation does not exist.
+This *only* applies to the `.into_iter()` method call syntax.
+It does not affect any other syntax such as `for e in [1, 2, 3]` or `iter.zip([1, 2, 3])`.
+Those will start to work in *all* editions.
+
+While it's a shame that this required a small hack to avoid breakage,
+we're very happy with how this solution keeps the difference between
+the editions to an absolute minimum.
```

Comment:
```suggestion
the editions to an absolute minimum while restricting any
added complexity exclusively to older editions.
```
I've seen a lot of people confused about the scope of this hack, so adding a single line to reemphasize that the hack isn't observable in the new edition could be useful.